dora/underfs/hdfs/src/test/java/alluxio/underfs/hdfs/HdfsUnderFileSystemIntegrationTestBase.java

```diff
@@ -0,0 +1,146 @@
+    mHdfsConfiguration.set("dfs.data.dir", mTemp.newFolder("dn").getAbsolutePath());
+    // 1MB block size for testing to save memory
+    mHdfsConfiguration.setInt("dfs.block.size", BLOCK_SIZE);
+
+    mCluster = new MiniDFSCluster.Builder(mHdfsConfiguration)
+        .enableManagedDfsDirsRedundancy(false)
+        .manageDataDfsDirs(false)
+        .manageNameDfsDirs(false)
+        .numDataNodes(1).build();
+
+    UnderFileSystemConfiguration ufsConf =
+        UnderFileSystemConfiguration.defaults(alluxio.conf.Configuration.global());
+
+    setConfiguration();
+
+    mUfs = createUfs(ufsConf);
+  }
+
+  @After
+  public void after() {
+    if (mCluster != null) {
+      mCluster.close();
+    }
+    if (mUfs != null) {
+      mUfs.close();
+    }
+  }
+
+  protected void writeMultiBlockFileTest(String testFilePath) throws IOException {
+    // 16MB + 1 byte, 17 blocks
+    int fileLength = 1024 * 1024 * 16 + 1;
+    int numHdfsBlocks = (fileLength - 1) / BLOCK_SIZE + 1;
+
+    RandomStringGenerator randomStringGenerator =
+        new RandomStringGenerator.Builder()
+            .withinRange('0', 'z')
+            .filteredBy(CharacterPredicates.LETTERS, CharacterPredicates.DIGITS)
+            .build();
+    String fileContentToWrite = randomStringGenerator.generate(fileLength);
+
+    OutputStream os = mUfs.create(testFilePath, getCreateOption());
+    os.write(fileContentToWrite.getBytes());
+    os.close();
+
+    InputStream is = mUfs.open(testFilePath);
+    String readFileContent = IOUtils.toString(is);
+    Assert.assertEquals(fileContentToWrite, readFileContent);
+
+    assertEquals(fileLength, mUfs.getStatus(testFilePath).asUfsFileStatus().getContentLength());
+    FileStatus status = mUfs.getFs().getFileStatus(new Path(testFilePath));
+    assertEquals(numHdfsBlocks,
+        mUfs.getFs().getFileBlockLocations(status, 0, status.getLen()).length);
+  }
+
+  protected void writeEmptyFileTest() throws IOException {
+    String testFilePath = "/empty_file";
+    OutputStream os = mUfs.create(testFilePath, getCreateOption());
+    os.close();
+    assertEquals(0, mUfs.getStatus(testFilePath).asUfsFileStatus().getContentLength());
+  }
+
+  protected void hdfsDownDuringUploadTest() throws Exception {
+    String testFilePath = "/test_file";
+
+    RandomStringGenerator randomStringGenerator =
+        new RandomStringGenerator.Builder()
+            .withinRange('0', 'z')
+            .filteredBy(CharacterPredicates.LETTERS, CharacterPredicates.DIGITS)
+            .build();
+    String oneBlockFileContent = randomStringGenerator.generate(1024 * 1024);
+
+    OutputStream os = mUfs.create(testFilePath, getCreateOption());
+    os.write(oneBlockFileContent.getBytes());
+    os.write(oneBlockFileContent.getBytes());
+    // Stop the data node in the middle of the write process.
+    mCluster.stopDataNode(0);
+    os.write(oneBlockFileContent.getBytes());
+    os.close();
+  }
```

Comment:
But it does not verify the write result?
e.g. to check file existence, file length, file content?

Reply:
this one is used to test the case where the hdfs datanode is down during the write process. The user is expected to get an exception. The behavior on the HDFS side is not defined.

Reply:
The test above: writeMultiBlockFileTest verifies the result the stream wrote.